CT-2589 Move team members to new business unit

danielatdigital opened the pull request:
## Description
<!-- Description of the changes. High level overview of the requirements and the attempted solution -->
 
## Self-review checklist
<!-- Action these things before requesting reviews -->
* [ ] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [x] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers
 
### Screenshots
<!-- Screenshots of the new changes if appropriate -->
 
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2589?atlOrigin=eyJpIjoiMGZhY2Y3ZDIyYTAyNDkwMWE4ODBlZGMyZGI1MWQ3MzQiLCJwIjoiaiJ9
 
### Deployment
<!-- Notes about database migrations, new runtime dependencies, mitigating downtime, feature flags, etc -->
 
### Manual testing instructions
<!-- Step-by-step instructions for the reviewer to manually test the changes -->

Commit:
Add foreign key to self reference team
    Amend report to show moved team

    This is the preparation work for moving business units between directorate
    and the reporting on this

Commit:
Fixed error checking for nil
And referencing destination unit's parent name (Always a Directorate for a Business Unit)

Commit:
Add structure.sql to branch

It should have been there all along

Polsonby committed:
CT-2450 Disable unhelpful Bullet checking for STI association

Commit:
Amend code for preferred style

More idiomatic

danielatdigital committed:
Update team.rb

Removed duplicate code

Commit:
Initial Test

To check we're not moving the team to it's original directorate

Commit:
Move (duplicate) team, amend to new directorate, add meta data

Future tickets include deactivating previous team record, and adding foriegn ket to new team

Commit:
Add check to prove that target directory parameter is a directorate

Because teams class is inherited by Business Group, Directorate, and Business Unit
... and we don't want to move to be a child of Business Unit

Commit:
Squash rubocop warnings

To make code shiny

danielatdigital committed:
Merge branch 'master' into CT-2586-Create-cloned-business-unit

Commit:
Amend formatting

For prettier display

Commit:
Add team properties to the new team

For completeness

Commit:
Copy user_roles to new business unit

Remaining: Remove those roles from original unit

Commit:
Move team members to new team (removing them from the old team)
So that the new version of team has members in it

danielatdigital committed:
Merge branch 'master' into CT2589-Move-team-members-to-new-business-unit

Commit:
Initial Commit

mseedat-moj reviewed (changes requested):
General functionality is fine, some formatting suggestions, and a single general question:

- How is the Team Move Service guaranteed atomicity? e.g. transactions, flags etc and therefore how is failure handled? Is that for the next phase of work?

Atomicity handled in a later ticket

danielatdigital committed:
Amend formatting

as per review

Commit:
Merge branch 'CT2589-Move-team-members-to-new-business-unit' of github.com:ministryofjustice/correspondence_tool_staff into CT2589-Move-team-members-to-new-business-unit

Commit:
Add method to Business Unit
To allow destruction of roles - the roles are copied to the new team prior to destruction

Commit:
Amend service to use initialise/call pattern
To match the style of other services in the application

danielatdigital committed:
Merge branch 'master' into CT2589-Move-team-members-to-new-business-unit

Commit:
CT-2589 Review updates (recommended)

mseedat-moj reviewed (approved):
Good job, thanks for the changes:

Had a telephone call to discuss the following:

- TeamMoveService.call transaction should be _inside_ the try/catch block
- move_team would not have thrown an exception if save failed, use `.save!`
- renamed `moved_team` to `move_team!` to reflect exceptional nature

danielatdigital committed:
Add the ticket number to the TODO
As per review

danielatdigital committed:
Merge branch 'master' into CT2589-Move-team-members-to-new-business-unit

danielatdigital committed:
Merge branch 'master' into CT2589-Move-team-members-to-new-business-unit

danielatdigital committed:
Merge branch 'master' into CT2589-Move-team-members-to-new-business-unit
